### PR TITLE
Annotate SyntaxNodeOrToken

### PR DESCRIPTION
**Pull request by jaredpar**

### CODE REVIEW

**Comment by jaredpar** [comment]
@roslyn-compiler PTAL

**Comment by jaredpar** [comment]
@dotnet/roslyn-compiler  PTAL

File: src/Compilers/Core/Portable/Syntax/SyntaxNodeOrToken.cs [hunk 1]
```diff
@@ -174,14 +189,29 @@ public SyntaxNode AsNode()
             return _nodeOrParent;
         }
 
+        internal bool AsNode([NotNullWhen(true)] out SyntaxNode? node)
+        {
+            if (IsNode)
+            {
+                node = _nodeOrParent;
+                return node is object;
```
Comment (cston) [inline src/Compilers/Core/Portable/Syntax/SyntaxNodeOrToken.cs]: >return node is object; [](start = 16, length = 22)

It looks like the callers of `AsNode(out SyntaxNode)` where previously using the result of `AsNode()` if `IsNode` returned `true`. Now those callers are skipping the case where `IsNode == true` but `_nodeOrParent == null`.

Is that behavior change observable and expected?

Comment (jaredpar) [reply src/Compilers/Core/Portable/Syntax/SyntaxNodeOrToken.cs]: That's true. This means really `SyntaxNodeOrToken` really has four states:
- It's a node
- It's a token
- It's default
- It's classified as a node without actually having a node

This type is plain weird. I'm not really sure how to reconcile this.

**Comment by jaredpar** [comment]
@333fred note I made one more commit which removed the fourth state of `SyntaxNodeOrToken`

**Pull request merged**